### ISSUE TITLE
bug: SSL_get_version returned TLSv1.3

### ISSUE THREAD

**Issue by copy** [janestreet/async_ssl#31]
I get the below failure when calling `Async_ssl.Ssl.Connection.version` on a tls 1.3 connection:

```
((2020-01-31 10:55:50.739216595+01:00)
 ("unhandled exception"
  ((monitor.ml.Error (Failure "bug: SSL_get_version returned TLSv1.3")
    ("Raised at file \"stdlib.ml\", line 29, characters 17-33"
     "Called from file \"httpcheck.ml\", line 332, characters 30-73"
     "Called from file \"async/private_ssl_real.ml\", line 199, characters 8-30"
     "Called from file \"src/deferred0.ml\", line 54, characters 64-69"
     "Called from file \"src/job_queue.ml\", line 167, characters 6-47"
     "Caught by monitor main"))
   ((pid 21165) (thread_id 0)))))
```

**Comment by tiash** [comment]
Async_ssl doesn't currently support TLSv1.3.

Adding support unfortunately requires subtle changes beyond just
extending the enumeration, and breaks compatibility with older
versions of the underlying openssl library.

**Comment by copy** [comment]
It seems to be using TLS 1.3 already (successfully, unless `Async_ssl.Ssl.Connection.version` is called; verified using wireshark):

```ocaml
open Core
open Async

let main () =
  let uri = Uri.of_string "https://google.com/" in
  let request = Cohttp.Request.make uri in
  let verify connection =
    let _version = Async_ssl.Ssl.Connection.version connection in
    return true
  in
  let ssl_config = Conduit_async.V2.Ssl.Config.create
      ~verify_modes:[Async_ssl.Verify_mode.Verify_peer]
      ~hostname:(Uri.host uri |> Option.value_exn)
      ~ca_file:"/usr/share/ca-certificates/trust-source/mozilla.trust.p11-kit"
      ~verify () in
  let%bind (r, _) = Cohttp_async.Client.request ~ssl_config ~uri request in
  Format.printf "%a\n%!" Cohttp.Response.pp_hum r;
  return ()

let () = don't_wait_for (main ()); Scheduler.go () |> never_returns
```

**Comment by aalekseyev** [comment]
This is supposed to be fixed in 2cc103c, so closing.

**Comment by anuragsoni** [comment]
@aalekseyev It seems like https://github.com/janestreet/async_ssl/commit/2cc103c0648f5eec8717cd6b09b48bfff21cf3e9 did not end up making it to the 0.15 version release to opam. I notice the same exception reported on this issue when calling `get_version` on a TLS 1.3 connection. Looking at the implementation on the 0.15 tag, the changes from the commit you linked to are indeed missing. https://github.com/janestreet/async_ssl/blob/03de97a8e380bc4c31b9fa0e67691eee23e76e0c/src/ffi__library_must_be_initialized.ml#L493-L503

**Comment by aalekseyev** [comment]
@aalekseyev, the way I'm reading it, commit `2cc103c` did not change the `get_version` function, and if you look at e.g. [Ssl_ctx.create_exn](https://github.com/janestreet/async_ssl/blob/03de97a8e380bc4c31b9fa0e67691eee23e76e0c/src/ffi__library_must_be_initialized.ml#L97), I can see the changes are actually preserved.
Maybe you're saying `get_version` also needs to be fixed? (@tiash, maybe you want to have a look?)

**Comment by anuragsoni** [comment]
You are completely correct! I missed that. But yeah, I think `get_version` also needs to be fixed. Thanks!

**Comment by tiash** [comment]
Yes, the original commit missed this.
`get_version` was fixed as part of [`5a3e7399`](https://github.com/janestreet/async_ssl/blame/86882526834e0d58a0e35b4ae4ce3303cb369bf6/src/ffi__library_must_be_initialized.ml#L485)

**Comment by anuragsoni** [comment]
Would it be possible to get a 0.15.1 release on opam with this patch on get_version?
